apps/query-ui/queryui/util.py

```diff
@@ -85,86 +143,136 @@ def docset_to_string(docset: DocSet) -> str:
-        width=700,
-        height=500,
-        directed=True,
-        physics=False,
-        hierarchical=True,
-        direction="UD",
-    )
-    agraph(nodes=nodes, edges=edges, config=config)
-
-
-@st.experimental_fragment
-def show_query_traces(trace_dir: str, query_id: str):
-    """Show the query traces in the given trace_dir."""
-    trace_dir = os.path.join(trace_dir, query_id)
-    for node_id in sorted(os.listdir(trace_dir)):
-        data_list = []
-        directory = os.path.join(trace_dir, node_id)
+class QueryNodeTrace:
+    """Display query traces for a single query node."""
+
+    # The order here is chosen to ensure that the most important columns are shown first.
+    # The logic below ensure that additional columns are also shown, and that empty columns
+    # are omitted.
+    COLUMNS = [
+        "properties.path",
+        "properties.entity.accidentNumber",
+        "properties.key",
+        "properties.count",
+        "properties.entity.location",
+        "properties.entity.dateAndTime",
+        "properties.entity.aircraft",
+        "properties.entity.registration",
+        "properties.entity.conditions",
+        "properties.entity.windSpeed",
+        "properties.entity.visibility",
+        "properties.entity.lowestCeiling",
+        "properties.entity.injuries",
+        "properties.entity.aircraftDamage",
+        "text_representation",
+        "doc_id",
+        "parent_id",
+    ]
+
+    def __init__(self, trace_dir: str, node_id: str):
+        self.trace_dir = trace_dir
+        self.node_id = node_id
+        self.df = None
+        self.readdata()
+
+    def readfile(self, f):
+        with open(f, "rb") as file:
+            try:
+                doc = pickle.load(file)
+            except EOFError:
+                return None
+
+            # For now, skip over MetadataDocuments.
+            if "metadata" in doc.keys():
+                return None
+
+            # Flatten properties.
+            if "properties" in doc:
+                for prop in doc["properties"]:
+                    if isinstance(doc["properties"][prop], dict):
+                        for nested_property in doc["properties"][prop]:
+                            doc[".".join(["properties", prop, nested_property])] = doc["properties"][prop][
+                                nested_property
+                            ]
+                    else:
+                        doc[".".join(["properties", prop])] = doc["properties"][prop]
+                doc.pop("properties")
+
+            # Keep only the columns we care about.
+            # Limit size of each to avoid blowing out memory.
+            def format_value(value):
+                MAX_CHARS = 1024
+                if isinstance(value, str):
+                    if len(value) > MAX_CHARS:
+                        return value[:MAX_CHARS] + "..."
+                    else:
+                        return value
+                if isinstance(value, dict):
```

Comment:
I would think you would want return format_value(str(value)). If you end up with a big class in the value it seems like you would want to truncate it.
I'd probably then just rewrite to:
```
  if not isinstance(value, str):
       value = str(value)
  MAX_CHARS = 1024 
  if len(value) > MAX_CHARS:
       return value[:MAX_CHARS] + "..."
  else:
       return value
```